edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs

```diff
@@ -430,15 +458,17 @@ async Task HandleNoServiceIdentity(string id)
         async Task HandleNewServiceIdentity(ServiceIdentity serviceIdentity)
         {
             Option<ServiceIdentity> existing = await this.serviceIdentityHierarchy.Get(serviceIdentity.Id);
-            bool hasUpdated = existing.HasValue && !existing.Contains(serviceIdentity);
-
-            await this.serviceIdentityHierarchy.InsertOrUpdate(serviceIdentity);
-            await this.SaveServiceIdentityToStore(serviceIdentity.Id, new StoredServiceIdentity(serviceIdentity));
-            Events.AddInScope(serviceIdentity.Id);
 
+            bool hasUpdated = await this.serviceIdentityHierarchy.AddOrUpdate(serviceIdentity);
             if (hasUpdated)
             {
-                this.ServiceIdentityUpdated?.Invoke(this, serviceIdentity);
+                Events.AddInScope(serviceIdentity.Id);
+                await this.SaveServiceIdentityToStore(serviceIdentity.Id, new StoredServiceIdentity(serviceIdentity));
+
+                if (existing.HasValue)
```

Comment:
>existing.HasValue [](http://example.com/codeflow?start=20&length=17)

shouldn't this now be (existing.HasValue && hasUpdated)?

Reply:
if it got to this line hasUpdated is true (because of line 463 if)